packages/stream-plugin/src/index.ts

```diff
@@ -44,46 +45,67 @@ export interface ActionCtx<T extends { [K in keyof T]: Stream<any> }> {
   [valueSymbol]: Partial<UnStreams<T>>;
 }
 
-const init = <T extends { [K in keyof T]: Stream<any> }>(
-  _config: {},
-  universe: Universe<T>,
-) => {
-  universe.streams = {};
-};
+export interface Event<T extends { [K in keyof T]: Stream<any> }> {
+  streams: { [K in keyof T]?: T[K] | (typeof deletedSymbol) };
+}
 
-const prepareActionCtx = <T extends { [K in keyof T]: Stream<any> }>(
-  state: State<T>,
-  streamUpdate: PluginAction<ActionCtx<T>, [keyof T, any]>,
-) => ({
+const prepareActionCtx = <T extends { [K in keyof T]: Stream<any> }>({
   ctx,
   universe,
+  event,
 }: {
-  ctx: PluginActionCtx<ActionCtx<T>, Universe<T>> & ActionCtx<T>;
+  ctx: ActionCtx<T>;
   universe: Universe<T>;
+  event: Event<T>;
 }) => {
+  // For the streamUpdate action
   ctx[valueSymbol] = universe.streams;
-  ctx.streams = new Proxy(state.streams, {
-    set: (obj, key: keyof T, value) => {
-      const cb = (value: any) => {
-        ctx.rootDispatch(streamUpdate)(key, value);
-      };
-      state.states[key] = value.subscribe(cb);
+
+  // For user actions
+  event.streams = {};
+  ctx.streams = new Proxy(event.streams as Partial<T>, {
+    get: () => undefined,
+    set: <K extends keyof T>(obj, key: K, value: T[K]) => {
       obj[key] = value;
       return true;
     },
     deleteProperty: (obj, key: keyof T) => {
-      const streamState = state.states[key];
-      if (streamState != null) {
-        streamState.unsubscribe();
-      }
+      // Typescript doesn't know about proxies (typescript#20846)
+      obj[key] = deletedSymbol as any;
       delete universe.streams[key];
-      delete state.states[key];
-      delete obj[key];
       return true;
     },
   });
 };
 
+const onCompleteEvent = <T extends { [K in keyof T]: Stream<any> }>(
+  state: State<T>,
+  streamUpdate: PluginAction<ActionCtx<T>, [keyof T, any]>,
+): PluginOnCompleteEventFn<{}, Event<T>, ActionCtx<T>> => ({
+  event,
+  rootDispatch,
+}) => {
+  // Not sure how to type the Object.entries version
+  (Object.keys(event.streams) as Array<keyof T>).forEach(key => {
```

Comment:
Strange that the type of `Object.keys` cannot be inferred :/

Reply:
It infers it as `string`. Same with entries, you get `[string, unknown][]`. There is some discussion [here](https://github.com/microsoft/TypeScript/pull/20553), the main issue being you can assign `{foo: number, bar: string}` to a `{foo: number}`, and then when you iterate over entries you get them all.

Reply:
The main issue is that a `{foo: number, bar: string}` can be assigned to a `{foo: number}`, so `Object.entries`/`Object.keys` cannot infer this. See [here](https://github.com/microsoft/TypeScript/pull/20553#issuecomment-350226895) for discussion.

Reply:
It's because a `{foo: number, bar: string}` can be assigned to a `{foo: number}`, see discussion [here](https://github.com/microsoft/TypeScript/pull/20553#issuecomment-350226895).